apiserver/common/credentialcommon/modelcredential.go

```diff
@@ -76,22 +77,33 @@ func checkCAASModelCredential(brokerParams environs.OpenParams) (params.ErrorRes
 	return params.ErrorResults{}, nil
 }
 
-func checkIAASModelCredential(openParams environs.OpenParams, backend PersistentBackend, callCtx context.ProviderCallContext, checkCloudInstances bool) (params.ErrorResults, error) {
+func checkIAASModelCredential(openParams environs.OpenParams, backend PersistentBackend, callCtx context.ProviderCallContext, checkCloudInstances bool, modelMigrationCheck bool) (params.ErrorResults, error) {
 	env, err := newEnv(callCtx, openParams)
 	if err != nil {
 		return params.ErrorResults{}, errors.Trace(err)
 	}
+
+	// Check that we can see all machines' instances regardless of their state as perceived by the
+	// cloud, i.e. this call will return all non-terminated instances.
+	instances, err := env.AllInstances(callCtx)
+	// If we're not performing this check for model migrations; then being able to get the instances
+	// is proof enough that the credential is valid (authenticated, authorization is a different
+	// concern), no need to check the mapping between instances and machines.
+	if !modelMigrationCheck || err != nil {
```

Comment:
If the error is non-nil, you _always_ have to return a populated error results.
We only return empty for non-nil _and_ !modelMigrationCheck.

Reply:
Yeah this is accurate. Though, in my defense, currently (before this change) the `fail` function here doesn't actually put the error into the ErrorResults https://github.com/juju/juju/blob/4ff4e14605426e323d1326810600c990de5dc427/apiserver/common/credentialcommon/modelcredential.go#L102